src/message_conversions.rs

```diff
@@ -0,0 +1,500 @@
+use protobuf::Enum;
+use std::time::Duration;
+use up_rust::{UCode, UMessage, UMessageBuilder, UMessageType, UPayloadFormat, UStatus, UUri};
+use vsomeip_sys::glue::{
+    make_message_wrapper, make_payload_wrapper, ApplicationWrapper, MessageWrapper, RuntimeWrapper,
+};
+use vsomeip_sys::safe_glue::{
+    get_data_safe, get_message_payload, get_pinned_application, get_pinned_message_base,
+    get_pinned_payload, get_pinned_runtime, offer_single_event_safe, set_data_safe,
+    set_message_payload,
+};
+use vsomeip_sys::vsomeip;
+use vsomeip_sys::vsomeip::{message_type_e, ANY_MAJOR};
+
+const UP_CLIENT_VSOMEIP_FN_TAG_CONVERT_UMSG_TO_VSOMEIP_MSG: &str = "convert_umsg_to_vsomeip_msg";
+const UP_CLIENT_VSOMEIP_FN_TAG_CONVERT_VSOMEIP_MSG_TO_UMSG: &str = "convert_vsomeip_msg_to_umsg";
+
+pub async fn convert_umsg_to_vsomeip_msg_and_send(
+    umsg: &UMessage,
+    application_wrapper: &mut UniquePtr<ApplicationWrapper>,
+    runtime_wrapper: &UniquePtr<RuntimeWrapper>,
+) -> Result<(), UStatus> {
+    let Some(source) = umsg.attributes.source.as_ref() else {
+        return Err(UStatus::fail_with_code(
+            UCode::INVALID_ARGUMENT,
+            "Message has no source UUri",
+        ));
+    };
+
+    match umsg
+        .attributes
+        .type_
+        .enum_value_or(UMessageType::UMESSAGE_TYPE_UNSPECIFIED)
+    {
+        UMessageType::UMESSAGE_TYPE_PUBLISH => {
+            let vsomeip_msg =
+                make_message_wrapper(get_pinned_runtime(runtime_wrapper).create_notification(true));
+            let (_instance_id, service_id) = split_u32_to_u16(source.ue_id);
+            get_pinned_message_base(&vsomeip_msg).set_service(service_id);
+            let instance_id = 1; // TODO: Setting to 1 manually for now
+            get_pinned_message_base(&vsomeip_msg).set_instance(instance_id);
+            let (_, event_id) = split_u32_to_u16(source.resource_id);
+            get_pinned_message_base(&vsomeip_msg).set_method(event_id);
+            let client_id = 0; // already set in underlying vsomeip code, but we ensure it's set
+            get_pinned_message_base(&vsomeip_msg).set_client(client_id);
+            let (_, _, _, interface_version) = split_u32_to_u8(source.ue_version_major);
+            trace!("uProtocol Publish message's interface_version: {interface_version}");
+            get_pinned_message_base(&vsomeip_msg).set_interface_version(interface_version);
+            let payload = {
+                if let Some(bytes) = umsg.payload.clone() {
+                    bytes.to_vec()
+                } else {
+                    Vec::new()
+                }
+            };
+            let vsomeip_payload =
+                make_payload_wrapper(get_pinned_runtime(runtime_wrapper).create_payload());
+            set_data_safe(get_pinned_payload(&vsomeip_payload), &payload);
+            let payload = vsomeip_payload.get_shared_ptr();
+            // set_message_payload(&mut vsomeip_msg, &mut vsomeip_payload);
```

Comment:
is this needed anymore?

Reply:
It is not! Thanks for catching it.

vsomeip for their version of Publish (called "notify") as you can see below we just supply the payload.

There's actually a fair bit I can remove up above now that I think of it. :+1:

Reply:
Resolved this one